Fix building polymake on 10.10 and 10.7

catap opened the pull request:
#### Description

Here I've fixed building on 10.10 `scip` and on 10.7 `mumps` and `OpenBLAS`.

I've also revbumped `scalapack`.

And make explicit requirements of compiler for `polymake`.

All of it allows to build `polymake` on both target platforms.

Closes: https://trac.macports.org/ticket/67322
Closes: https://trac.macports.org/ticket/67175
Closes: https://trac.macports.org/ticket/67350

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [x] bugfix

###### Tested on
<!-- Triple-click and copy the next line and paste it into your shell. It will copy your OS and Xcode version to the clipboard. Paste it here replacing this section.
sh -c 'echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion) $(uname -m)"; xcode=$(xcodebuild -version 2>/dev/null); if [ $? == 0 ]; then echo "$(echo "$xcode" | awk '\''NR==1{x=$0}END{print x" "$NF}'\'')"; else echo "Command Line Tools $(pkgutil --pkg-info=com.apple.pkg.CLTools_Executables | awk '\''/version:/ {print $2}'\'')"; fi' | tee /dev/tty | pbcopy
-->

macOS 10.10.5 14F2511 x86_64
Xcode 7.2.1 7C1002

macOS 10.7.5 11G63 x86_64
Xcode 4.6.3 4H1503


###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? <!-- Please don't open a new Trac ticket if you are submitting a pull request. -->
- [x] checked your Portfile with `port lint --nitpick`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

catap commented:
@ryandesign here it is

ryandesign commented on math/OpenBLAS/Portfile:
```diff
@@ -6,11 +6,12 @@ PortGroup           muniversal 1.1
 PortGroup           compilers 1.0
 PortGroup           xcode_workaround 1.0
 PortGroup           makefile 1.0
+PortGroup           compiler_blacklist_versions 1.0
```
You're not using this portgroup.
```suggestion
```

ryandesign commented on math/OpenBLAS/Portfile:
```diff
@@ -19,6 +20,9 @@ compiler.cxx_standard 2011
 compiler.thread_local_storage yes
 
+# error: invalid instruction mnemonic 'cvtsi2ssl'
+compiler.blacklist-append {macports-clang-15}
```
I'm surprised that this would be the fix for this issue. Can you add a URL with a bug report or more information confirming that this error is a bug in clang 15? The Trac ticket only talks about the `mktemp` problem.

Curly braces are not required here.

catap replied:
@ryandesign I haven't find any bug report. I simple tried another clang (14) and it works :)

Yes, I've reproduced it on my machine before fixing it.

ryandesign commented on math/mumps/Portfile:
```diff
@@ -6,6 +6,7 @@ PortGroup                   linear_algebra 1.0
 PortGroup                   muniversal 1.0
 PortGroup                   makefile 1.0
 PortGroup                   conflicts_build 1.0
+PortGroup                   compiler_blacklist_versions 1.0
```
Same:
```suggestion
```

ryandesign commented on math/mumps/Portfile:
```diff
@@ -52,6 +53,9 @@ configure.optflags          -O3
 configure.cflags-append     -fPIC
 configure.fflags-append     -fPIC
 
+# error: invalid instruction mnemonic 'cvtsi2ssl'
+compiler.blacklist-append   {macports-clang-15}
```
Same.

ryandesign commented on math/polymake/Portfile:
```diff
@@ -20,6 +20,9 @@ checksums               rmd160  be170dff2b2838c9b65d48b08f4b679f60bbd670 \
                         size    47480264
 
+compiler.cxx_standard   2014
+compiler.c_standard     2014
```
There's no such thing as C 2014.

ryandesign commented on math/scip/Portfile:
```diff
@@ -27,6 +27,11 @@ checksums           rmd160  faa34bb47c67cdb7d5789de2e59fc8b09c1cd1f1 \
 
 compilers.setup     require_fortran
 
+# Block compilers that do not support thread-local storage.
+compiler.cxx_standard 2011
+compiler.c_standard   2011
+compiler.thread_local_storage yes
```
The comment doesn't seem necessary since the code is self-explanatory.

Looking at the `-std` flags used during the build, it seems to be using C 1999 and C++ 2014, doesn't it?

I already added `compiler.thread_local_storage yes` below.

```suggestion
compiler.c_standard   1999
compiler.cxx_standard 2014
```

ryandesign commented on math/OpenBLAS/files/patch-MacOSX-mktemp.diff:
```diff
@@ -1,6 +1,17 @@
+ 
+ no_avx512=0
+ if [ "$architecture" = "x86" ] || [ "$architecture" = "x86_64" ]; then
+-    tmpd=`mktemp -d`
++    tmpd=`gmktemp -d`
```
Can you add a URL at the top of the patchfile to the bug report about this problem? Upstream should be correcting this by using `mktemp` in a portable manner, rather than in a GNU-specific way. We don't want to maintain this patch forever.

catap replied:
I haven't filed any issue, I simple adjusted the patch to the actual version.

catap replied:
Backported to upstream

catap commented:
Trying to make something cleaner that blacklist of clang-15, or at least understand wtf

catap commented:
I figured it out. Clang-15 isn't a problem, but once I blacklisted it, I covered a real problem.

After I started testing each port by building as `sudo port -f deactivate installed && sudo port -N install --unrequested XXX` to get a clean system, I've caught the root cause and fixed it.

The cause is a mix between `clang`, `gfortan` and `cctools`. The last one doesn't know anything about `llvm-15` and it leads to quite strange behavior that was reported.

My blacklisting of `clang-15` simply switched it to `clang-14` / `llvm-14` which is supported by `cctools` and everything works.

Once `llvm-14` is installed, it works fine with `clang-15` as the compiler. Once I've added support for `llvm-15` in `cctools`, it works fine without `llvm-14` :)

To avoid such problems in the future, I've added a note that `cctools` should also be changed if `compilers-1.0.tcl` is touched.

Mumps is a different story. It doesn't require a specific compiler version, and it can use a system one. So, it may faces the issue with cause similar with `OpenBLAS` If it goes by road with `clang-15`, or similar with `fpm`. So, I've reused the blacklist condition from that port here.

The only bad news is that `cctools` is currently broken and this PR is blocked by https://github.com/macports/macports-ports/pull/18443 where I'm fixing it.

kencu commented:
thank you for taking the time to find the "real" problem.

It is far too common to see people do a local "worked for me" fix without doing the proper analysis. That usually just leads to more trouble down the road as someone runs into trouble again later, and has to undo the previous fixes to finally get it properly sorted out.

I know people get frustrated when the "Well, it fixed it for me!" solution is not committed. But doing the proper analysis pays off in the end.

catap commented:
It is ready for review. @ryandesign, @kencu ?

kencu commented:
your cmake fix didn't actually add clang-16, but the addition of clang-15 looks fine to me.

catap commented:
@kencu thanks to catch that.

catap commented:
Added link to ticket https://trac.macports.org/ticket/67350 which should be fixed by this PR

catap commented:
@mascguy may I ask you to review and probably merge this one? Thanks!

catap commented:
Thanks, Christopher!
